### ISSUE TITLE
Require that physical "bits" be collected to support arithmetic operations of a given bitwidth

### ISSUE THREAD

**Issue by kostmo** [swarm-game/swarm#1229]
This is a potential mechanic for either challenge scenarios or classic mode.

If one wishes to compute the number 8, then they must have at least 3 bits equipped.
A special device may bypass this requirement.

This check would be performed in [`evalArith`](https://github.com/swarm-game/swarm/blob/e2b4eaa1dad3fe560342dcd721171b7f9fa2d2ac/src/Swarm/Game/Step.hs#L2660).  Values larger than the current bit width would throw an integer overflow exception.

**Comment by byorgey** [comment]
Hah, I love it!  You wouldn't know it, but over the year or two before starting work on Swarm I went through several different designs for some kind of robot programming/exploration game; one of my ideas was for something like this where bits are a physical resource that you have to mine to do operations.  Like if you wanted to increment a value you needed to use up a 1 bit.

Something like what you propose would certainly be fun as a challenge scenario mechanic. I am a little concerned it might be too annoying in classic mode but I think it could work well.  Bits are plentiful and you can get pretty far before you need to start computing things with numbers.

**Comment by byorgey** [comment]
Here's a concrete proposal.  Currently, a `calculator` device allows doing any arithmetic operations, and requires a `counter` which is constructed from a `flash memory`, which in turn requires 16 bits (8 0's and 8 1's).  We could consider making variants of this, e.g. an 8-bit flash memory, 16-bit, and 32-bit, with corresponding counter and calculator variants.  An n-bit calculator could allow doing only n-bit operations.  Perhaps there should also be a way to construct a calculator with unlimited bit width (like the current `calculator`).

**Comment by byorgey** [comment]
In fact, I kind of like the idea of NOT having a calculator with unlimited bit width - you would have to program your own arbitrary-precision operations.

Along with that, I wonder about making the `Int` type have a limited number of bits as well.  I think I will open another issue to start a discussion about that.
